File: app/src/main/java/tw/y_studio/ptt/UI/DragItemMove/ItemMoveCallback.java
package tw.y_studio.ptt.UI.DragItemMove;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.RecyclerView;


import tw.y_studio.ptt.Adapter.FavoriteBoardsListAdapter;

public class ItemMoveCallback extends ItemTouchHelper.Callback {

    private final ItemTouchHelperContract mAdapter;

    public ItemMoveCallback(ItemTouchHelperContract adapter) {
        mAdapter = adapter;
    }

    @Override
    public boolean isLongPressDragEnabled() {
        return false;
    }

    @Override
    public boolean isItemViewSwipeEnabled() {
        return false;
    }



    @Override
    public void onSwiped(@NonNull RecyclerView.ViewHolder viewHolder, int i) {

    }

    @Override
    public int getMovementFlags(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder) {
        int dragFlags = ItemTouchHelper.UP | ItemTouchHelper.DOWN;
        return makeMovementFlags(dragFlags, 0);
    }

    private void changeTag(RecyclerView.ViewHolder viewHolder,int position){
        if(viewHolder instanceof FavoriteBoardsListAdapter.ViewHolderEdit){
            FavoriteBoardsListAdapter.ViewHolderEdit view = (FavoriteBoardsListAdapter.ViewHolderEdit) viewHolder;
            view.unfav.setTag(position);
        }
    }
    @Override
    public boolean onMove(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder,
                          RecyclerView.ViewHolder target) {
        viewHolder.itemView.setTag(target.getAdapterPosition());
        target.itemView.setTag(viewHolder.getAdapterPosition());
        //changeTag(viewHolder,target.getAdapterPosition());
        //changeTag(target,viewHolder.getAdapterPosition());

        mAdapter.onRowMoved(viewHolder.getAdapterPosition(), target.getAdapterPosition());
        //viewHolder.itemView.setTag(target.getAdapterPosition());
        //target.itemView.setTag(viewHolder.getAdapterPosition());
        return true;
    }

    @Override
    public void onSelectedChanged(RecyclerView.ViewHolder viewHolder,
                                  int actionState) {


        if (actionState != ItemTouchHelper.ACTION_STATE_IDLE) {
            if (viewHolder instanceof FavoriteBoardsListAdapter.ViewHolderEdit) {
                FavoriteBoardsListAdapter.ViewHolderEdit myViewHolder=
                        (FavoriteBoardsListAdapter.ViewHolderEdit) viewHolder;
                mAdapter.onRowSelected(myViewHolder);
            }

        }

        super.onSelectedChanged(viewHolder, actionState);
    }
    @Override
    public void clearView(RecyclerView recyclerView,
                          RecyclerView.ViewHolder viewHolder) {
        super.clearView(recyclerView, viewHolder);

        if (viewHolder instanceof FavoriteBoardsListAdapter.ViewHolderEdit) {
            FavoriteBoardsListAdapter.ViewHolderEdit myViewHolder=
                    (FavoriteBoardsListAdapter.ViewHolderEdit) viewHolder;
            mAdapter.onRowClear(myViewHolder);
        }
    }

    public interface ItemTouchHelperContract {

        void onRowMoved(int fromPosition, int toPosition);
        void onRowSelected(FavoriteBoardsListAdapter.ViewHolderEdit myViewHolder);
        void onRowClear(FavoriteBoardsListAdapter.ViewHolderEdit myViewHolder);

    }


}

File: app/src/main/java/cc/ptt/android/presentation/common/dragitemmove/ItemMoveCallback.kt
package cc.ptt.android.presentation.common.dragitemmove

import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView

class ItemMoveCallback(
    private val helperContract: ItemTouchHelperContract
) : ItemTouchHelper.Callback() {
    override fun isLongPressDragEnabled(): Boolean {
        return false
    }

    override fun isItemViewSwipeEnabled(): Boolean {
        return false
    }

    override fun onSwiped(viewHolder: RecyclerView.ViewHolder, i: Int) {
    }

    override fun getMovementFlags(
        recyclerView: RecyclerView,
        viewHolder: RecyclerView.ViewHolder
    ): Int {
        val dragFlags = ItemTouchHelper.UP or ItemTouchHelper.DOWN
        return makeMovementFlags(dragFlags, 0)
    }

    override fun onMove(
        recyclerView: RecyclerView,
        viewHolder: RecyclerView.ViewHolder,
        target: RecyclerView.ViewHolder
    ): Boolean {
        viewHolder.itemView.tag = target.adapterPosition
        target.itemView.tag = viewHolder.adapterPosition
        helperContract.onRowMoved(viewHolder.adapterPosition, target.adapterPosition)
        return true
    }

    override fun onSelectedChanged(viewHolder: RecyclerView.ViewHolder?, actionState: Int) {
        if (actionState != ItemTouchHelper.ACTION_STATE_IDLE) {
            helperContract.onRowSelected(viewHolder)
        }
        super.onSelectedChanged(viewHolder, actionState)
    }

    override fun clearView(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder) {
        super.clearView(recyclerView, viewHolder)
        helperContract.onRowClear(viewHolder)
    }

    interface ItemTouchHelperContract {
        fun onRowMoved(fromPosition: Int, toPosition: Int)
        fun onRowSelected(viewHolder: RecyclerView.ViewHolder?)
        fun onRowClear(viewHolder: RecyclerView.ViewHolder?)
    }
}
